### ISSUE TITLE
TScatter palette settings

### ISSUE THREAD

**Issue by ferdymercury** [root-project/root#15922]
### Explain what you would like to see improved and how.

In the new TScatter class, it was found here https://root-forum.cern.ch/t/scatter-plot-with-colorbar/59562/14 that:

- There is not an easy way to control the color bar range
- Save TScatter object as .C and reload it, then the palette axis title is lost. The function SavePrimitive should be amended.

### ROOT version

```
ROOT v6.32.00
Built for linuxx8664gcc on May 28 2024, 05:02:02
From tags/v6-32-00@v6-32-00
With c++ (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Binary directory: /opt/root/bin
```

### Installation method

Binary Release

### Operating system

Ubuntu 22.04

**Comment by ferdymercury** [comment]
Thanks for the fix @couet!

I would also suggest adding an alternative way (commented) in the tutorial to "zoom" rather than "setting user range", namely to use:

```
   scatter->GetHistogram()->SetMinimum(10);
   scatter->GetHistogram()->SetMaximum(200);
```

This way, the palette is also zoomed, unlike with SetUserRange, which just acts as a filter on what points to plot or not.

Also: there is a bug:
- If one calls SetMinimum only, then the plot is broken. If that's intended, then maybe we should create a function scatter->SetMinMax() so that people don't forget to put both.

- the behavior of the TH2F when zooming Z-axis is different from TScatter. In TH2F, colors outside the range are clipped to max or min color. Here in TScatter, instead of clipping, they are filtered out of the plot. I don't know if that's a good solution, but if it is intended, it should be documented in the Doxygen docu. If not inteded, then the color should be clipped to the max/min of the palette.

**Comment by ferdymercury** [comment]
Thanks a lot for addressing one of the points! I am reopening to keep track of two minor bugs:

> Also: there is a bug:
> 
>     * If one calls SetMinimum only, then the plot is broken. If that's intended, then maybe we should create a function scatter->SetMinMax() so that people don't forget to put both.
> 
>     * the behavior of the TH2F when zooming Z-axis is different from TScatter. In TH2F, colors outside the range are clipped to max or min color. Here in TScatter, instead of clipping, they are filtered out of the plot. I don't know if that's a good solution, but if it is intended, it should be documented in the Doxygen docu. If not inteded, then the color should be clipped to the max/min of the palette.

**Comment by couet** [comment]
> * If one calls SetMinimum only, then the plot is broken. If that's intended, then maybe we should create a function scatter->SetMinMax() so that people don't forget to put both.


It is understood. The internal limits along Z (Minimum and Maximum) for the fHistogram (internal histogram of TScatter) are defined automatically when the histogram is Painted (i.e., Draw and Update). At TScatter creation time, both min and max are set to 0. So if you set the Minimum before drawing, the fHistogram min and max are not yet properly computed, and therefore you will have a minimum greater than the maximum (min = 10 and max = 0), and that's why you get this wrong plot. So you should simply do:

```
scatter->Draw("A");
canvas->Update();
scatter->GetHistogram()->SetMinimum(10);
```
I agree we can maybe set min and max also at the first call of GetHistogram, but that needs to be checked closely because we have exactly the same issue with TGraph2D (I have a reproducer):

In TGraph2D::GetHistogram, the Min and Max along Z are computed only if the option "empty" is passed to GetHistogram. I am looking into it.

**Comment by couet** [comment]
> If one calls SetMinimum only, then the plot is broken. If that's intended, then maybe we should create a function scatter->SetMinMax() so that people don't forget to put both.

fix here : https://github.com/root-project/root/pull/19101

**Comment by couet** [comment]
For this one:

the behavior of the TH2F when zooming Z-axis is different from TScatter. In TH2F, colors outside the range are clipped to max or min color. Here in TScatter, instead of clipping, they are filtered out of the plot. I don't know if that's a good solution, but if it is intended, it should be documented in the Doxygen docu. If not inteded, then the color should be clipped to the max/min of the palette.


what do you mean by zoom ? along Z axis or along x and Y ?

**Comment by ferdymercury** [comment]
> what do you mean by zoom ? along Z axis or along x and Y ?

I meant zooming in the Z axis, in other words, if I set the minimum Z in the palette to 10, all points below 10 will disappear (or be transparent).

In contrast, with a TH2F, if I set the minimum Z in the palette to 10, all bins below 10 will have the bin color as if the data were 10 and not be transparent.

So this fact about "disappearing" in the TScatter class is a bit awkward, as if you were missing data.

**Comment by couet** [comment]
Ok I see, it is intented. A scatter plot is not an histogram points out of range on any of the 3 Axis (X Y or Z) should not be drawn.

Edit: BTW it does not arm to add this sentence in the doc similarly to the explanation done for the COL option in THistPainter.

**Comment by ferdymercury** [comment]
Thanks. But wait, this behavior is against of what matplotlib or matlab does for color scatter plots, right? Isn't that a bit counterintuitive? To disappear instead of color clip?

**Comment by ferdymercury** [comment]
Consider the following snippet:

```
import matplotlib.pyplot as plt

plt.figure()
xyc = range(20)
plt.scatter(xyc, xyc, c=xyc, s=35, vmin=0, vmax=20)
plt.colorbar()
plt.xlim(0, 20)
plt.ylim(0, 20)
plt.show()

plt.figure()
xyc = range(20)
plt.scatter(xyc, xyc, c=xyc, s=35, vmin=5, vmax=15)
plt.colorbar()
plt.xlim(0, 20)
plt.ylim(0, 20)
plt.show()
```

I think this is the usual behavior also in other plotting libraries. In the second example, the points are still drawn, just the color is clipped.

Or at least, I would suggest that in ROOT one should be able to specify via a Option_t whether ROOT will clip colors or just remove points if the color value is outside the palette limits.

**Comment by couet** [comment]
I will try your code. I agree that it should behave like the others.

**Comment by couet** [comment]
Yes, I see what you mean. I would say that the default could be what mathplotlib does and, as you suggested, a drawing option ("CLIPCOL" ?) would produce the plot as it is now. ?

**Comment by ferdymercury** [comment]
Thanks!
CLIP should mean restricting a value to a limit. So it should be named differently, sth like SKIP or so.

**Comment by couet** [comment]
I added the option CLIPCOL... So you suggest to change the name ? SKIPCOL  ?

**Comment by couet** [comment]
I changed to SKIPCOL https://github.com/root-project/root/pull/19101.

**Comment by couet** [comment]
Once that PR is merged I think we can close this issue ?

**Comment by ferdymercury** [comment]
Awesome, yes, it can be closed now! Thanks a lot.
